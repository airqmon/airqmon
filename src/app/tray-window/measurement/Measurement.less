.measurement__pane {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  margin-top: @base-spacing;
}

.measurement {
  flex-basis: 50%;
  flex-direction: row;
  margin-top: 8px;

  &__reading,
  &__description {
    text-align: center;
    font-weight: 300;
  }

  &__reading {
    font-size: @font-size + 1pt;
  }

  &__norm {
    font-size: @font-size - 2pt;
    font-weight: 500;

    &.treshold {
      &-0 {
        color: #theme.light.aq-index() [caqi-1];

        .dark-theme & {
          color: #theme.dark.aq-index() [caqi-1];
        }
      }
      &-1 {
        color: #theme.light.aq-index() [caqi-2];

        .dark-theme & {
          color: #theme.dark.aq-index() [caqi-2];
        }
      }
      &-2 {
        color: #theme.light.aq-index() [caqi-3];

        .dark-theme & {
          color: #theme.dark.aq-index() [caqi-3];
        }
      }
      &-3 {
        color: #theme.light.aq-index() [caqi-4];

        .dark-theme & {
          color: #theme.dark.aq-index() [caqi-4];
        }
      }
      &-4 {
        color: #theme.light.aq-index() [caqi-5];

        .dark-theme & {
          color: #theme.dark.aq-index() [caqi-5];
        }
      }
      &-5 {
        color: #theme.light.aq-index() [caqi-5];

        .dark-theme & {
          color: #theme.dark.aq-index() [caqi-5];
        }
      }
    }
  }

  &__unit {
    display: inline;

    &::before {
      content: '';
      white-space: pre;
    }
  }

  &__description {
    text-transform: uppercase;
    color: #theme.light.text.color() [secondary];
    font-size: @font-size - 2pt;

    .dark-theme & {
      color: #theme.dark.text.color() [secondary];
    }
  }
}
