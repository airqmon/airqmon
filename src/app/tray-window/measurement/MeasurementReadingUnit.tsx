import * as React from 'react';

export enum Unit {
  PM = 'PM',
  TEMP_C = 'TEMP_C',
  PRESSURE_PA = 'PRESSURE_PA',
  PERCENT = 'PERCENT',
}

interface IMeasurementReadingUnitProps {
  unit: Unit;
}

export const MeasurementReadingUnit: React.SFC<IMeasurementReadingUnitProps> = (props) => {
  const content = {
    [Unit.PM]: (
      <>
        Î¼g/m<sup>3</sup>
      </>
    ),
    [Unit.TEMP_C]: (
      <>
        <sup>&deg;</sup>C
      </>
    ),
    [Unit.PRESSURE_PA]: 'hPA',
    [Unit.PERCENT]: '%',
  }[props.unit];

  return <div className="measurement__unit"> {content}</div>;
};
