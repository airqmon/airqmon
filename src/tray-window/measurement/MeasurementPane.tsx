import * as React from 'react';

import { Contaminants } from '../../contamination';
import { formatters, Measurement } from './Measurement';
import { Unit } from './MeasurementReadingUnit';
import { MeasurementValue, MeasurementValueNames } from '../../airqmon-api';

interface IMeasurementPaneProps {
  measurement?: MeasurementValue[];
}

const MeasurementPane: React.SFC<IMeasurementPaneProps> = (props) => {
  const { measurement } = props;

  const getValue = (measurement, name: MeasurementValueNames): number | null => {
    return (measurement.find((value) => value.name == name) || {}).value || null;
  };

  const pressure = getValue(measurement, MeasurementValueNames.PRESSURE);

  return (
    <div className="measurement__pane">
      <Measurement
        contaminant={Contaminants.PM25}
        reading={getValue(measurement, MeasurementValueNames.PM25)}
        formatter={formatters.significant}
      />
      <Measurement
        contaminant={Contaminants.PM10}
        reading={getValue(measurement, MeasurementValueNames.PM10)}
        formatter={formatters.significant}
      />
      <Measurement
        contaminant={Contaminants.PM1}
        reading={getValue(measurement, MeasurementValueNames.PM1)}
        formatter={formatters.significant}
      />
      <Measurement
        description="Temperature"
        unit={Unit.TEMP_C}
        reading={getValue(measurement, MeasurementValueNames.TEMPERATURE)}
        formatter={formatters.toFixed2}
      />
      <Measurement
        description="Pressure"
        unit={Unit.PRESSURE_PA}
        reading={pressure != null ? pressure / 100 : pressure}
        formatter={formatters.toFixed1}
      />
      <Measurement
        description="Humidity"
        unit={Unit.PERCENT}
        reading={getValue(measurement, MeasurementValueNames.HUMIDITY)}
        formatter={formatters.toFixed2}
      />
    </div>
  );
};

export default MeasurementPane;
