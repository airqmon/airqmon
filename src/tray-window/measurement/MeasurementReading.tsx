import * as React from 'react';

import { Unit, MeasurementReadingUnit } from './MeasurementReadingUnit';

export interface IMeasurementReadingProps {
  reading?: number;
  formatter?: (val: number) => string;
  unit?: Unit;
}

export const MeasurementReading: React.SFC<IMeasurementReadingProps> = ({
  reading,
  formatter,
  unit,
  children,
}) => (
  <>
    <div className="measurement__reading">
      {reading != null ? (formatter ? formatter(reading) : reading) : 'n/a'}
      {reading && unit ? <MeasurementReadingUnit unit={unit} /> : null}
      {children}
    </div>
  </>
);
