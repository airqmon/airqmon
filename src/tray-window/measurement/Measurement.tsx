import * as React from 'react';

import {
  Contaminants,
  CONTAMINATION_NORM_VALUES,
  CONTAMINATION_THRESHOLDS,
  getContaminationThresholdIndex,
} from 'common/contamination';

import { MeasurementReading, IMeasurementReadingProps } from './MeasurementReading';
import { Unit } from './MeasurementReadingUnit';

export const formatters: { [key: string]: (val: number) => string } = {
  significant: (val) => val.toFixed(0),
  toFixed2: (val) => val.toFixed(2),
  toFixed1: (val) => val.toFixed(1),
};

const CONTAMINANT_DESCRIPTIONS = {
  [Contaminants.PM1]: 'PM1',
  [Contaminants.PM10]: 'PM10',
  [Contaminants.PM25]: 'PM2.5',
};

const CONTAMINANT_UNITS = {
  [Contaminants.PM1]: Unit.PM,
  [Contaminants.PM10]: Unit.PM,
  [Contaminants.PM25]: Unit.PM,
};

const CONTAMINANT_FORMATTERS = {
  [Contaminants.PM1]: formatters.significant,
  [Contaminants.PM10]: formatters.significant,
  [Contaminants.PM25]: formatters.significant,
};

export interface IMeasurementProps extends IMeasurementReadingProps {
  contaminant?: Contaminants;
  description?: string;
  norm?: number;
}

export const Measurement: React.SFC<IMeasurementProps> = ({
  contaminant,
  reading,
  description = CONTAMINANT_DESCRIPTIONS[contaminant],
  formatter = CONTAMINANT_FORMATTERS[contaminant],
  norm = CONTAMINATION_NORM_VALUES[contaminant],
  unit = CONTAMINANT_UNITS[contaminant],
}) => {
  let normContent: JSX.Element = null;
  if (contaminant && CONTAMINATION_THRESHOLDS[contaminant]) {
    const normContentClassName = `measurement__norm treshold-${getContaminationThresholdIndex(
      contaminant,
      reading,
    )}`;
    normContent =
      reading && norm ? (
        <div className={normContentClassName}>{((reading / norm) * 100).toFixed(0)}%</div>
      ) : null;
  }

  return (
    <div className="measurement">
      <MeasurementReading reading={reading} unit={unit} formatter={formatter}>
        {normContent}
      </MeasurementReading>
      <div className="measurement__description">{description}</div>
    </div>
  );
};
