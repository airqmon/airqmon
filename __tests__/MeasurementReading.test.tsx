import * as React from 'react';
import { shallow } from 'enzyme';
import { MeasurementReading } from '../src/app/tray-window/measurement/MeasurementReading';
import { Unit } from '../src/app/tray-window/measurement/MeasurementReadingUnit';

jest.mock('../src/app/tray-window/measurement/MeasurementReadingUnit', () => ({
  ...require.requireActual('../src/app/tray-window/measurement/MeasurementReadingUnit'),
  MeasurementReadingUnit: 'MeasurementReadingUnit',
}));

describe('MeasurementReading', () => {
  const reading = 1.23;
  const formatter = jest.fn((val) => val.toFixed(0));

  afterEach(() => {
    formatter.mockClear();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<MeasurementReading reading={reading} formatter={formatter} />);
    expect(wrapper).toMatchSnapshot();
    expect(formatter).toHaveBeenCalledTimes(1);
    expect(formatter).toHaveBeenCalledWith(reading);
    expect(wrapper.childAt(0).hasClass('measurement__reading')).toBe(true);
  });

  it('renders correctly without formatter', () => {
    const wrapper = shallow(<MeasurementReading reading={reading} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders passed children', () => {
    const wrapper = shallow(
      <MeasurementReading reading={reading}>
        <div>A children node</div>
      </MeasurementReading>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('Renders MeasurementReadingUnit if unit is provided', () => {
    const wrapper = shallow(<MeasurementReading reading={reading} unit={Unit.PRESSURE_PA} />);
    expect(wrapper).toMatchSnapshot();
  });
});
