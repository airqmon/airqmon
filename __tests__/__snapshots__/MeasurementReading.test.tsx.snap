// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeasurementReading Renders MeasurementReadingUnit if unit is provided 1`] = `
<Fragment>
  <div
    className="measurement__reading"
  >
    1.23
    <MeasurementReadingUnit
      unit="PRESSURE_PA"
    />
  </div>
</Fragment>
`;

exports[`MeasurementReading renders correctly 1`] = `
<Fragment>
  <div
    className="measurement__reading"
  >
    1
  </div>
</Fragment>
`;

exports[`MeasurementReading renders correctly without formatter 1`] = `
<Fragment>
  <div
    className="measurement__reading"
  >
    1.23
  </div>
</Fragment>
`;

exports[`MeasurementReading renders passed children 1`] = `
<Fragment>
  <div
    className="measurement__reading"
  >
    1.23
    <div>
      A children node
    </div>
  </div>
</Fragment>
`;
